Issue #4800 - fix WebSocket LinkageError and invalid PathParam type handling

lachlan-roberts opened the pull request:
**Issue #4800**

Changes to the WebSocket MethodHandle lookup were needed as the implementation has changed in JDK 14 and our method of using the `MethodHandles.publicLookup()` no longer works.

The type of PathParams in the endpoint signature are now being checked at deployment, where they were previously being checked during the first upgrade request to that endpoint.

lachlan-roberts committed:
Change the WSServer to allow deployment of multiple webapps

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Issue #4800 - Reproduce LinkageError in WebSocket unit test

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Issue #4800 - prevent LinkageErrors for WS lookup on JDK 14

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Issue #4800 - invalid PathParam types should be reported at Deployment

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Issue #4800 - test both cases individually

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

olamy commented on tests/test-distribution/src/test/java/org/eclipse/jetty/tests/distribution/DistributionTests.java:
```diff
@@ -440,19 +439,14 @@ public void testSimpleWebAppWithWebsocket(String arg) throws Exception
-                assertEquals(HttpStatus.OK_200, response.getStatus());
-                assertThat(response.getContentAsString(), containsString("Hello"));
-                assertThat(response.getContentAsString(), not(containsString("<%")));
+                assertEquals(HttpStatus.SERVICE_UNAVAILABLE_503, response.getStatus());
```
I had the idea about adding another webapp to be sure it's still deployed despite having this LinkageError

lachlan-roberts replied:
The MethodHandle changes in this PR fix the LinkageError, but now this particular webapp should fail at deployment time anyway as the signature is wrong.

lachlan-roberts replied:
Should I add in another webapp here to test that it deploys successfully even if these two fail?

olamy replied:
yes. I would even say 2 :). I'm not sure about deployment order of contexts. maybe one first and one last. to ensure the failure do not break anything else to be deployed.

olamy replied:
@lachlan-roberts great tests changes! we now even have real websocket tests in the distribution tests module!

lachlan-roberts committed:
Issue #4800 - Improve WS DistributionTest to also test working webapp

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts commented:
Doesn't look like this fix for the MethodHandle.Lookup is going to work after all.

lachlan-roberts commented:
@sbordet the method where we use `MethodHandles.lookup()` and drop all lookupModes except PUBLIC does work fine on JDK 14, however it doesn't work for JDK 11 if done from a named module. All access is dropped when teleporting from a named module to another module except if you are using the publicLookup. So our resulting Lookup after the `Lookup.in()` has no access permissions at all.

So the  `publicLookup()` method only works before JDK14, and `MethodHandles.lookup()` + `dropLookupModes()` only works on JDK14.

lachlan-roberts committed:
Issue #4800 - Revert MethodHandle changes & update DistributionTests

- It seems it is correct to use the `MethodHandles.publicLookup().in()`
but this does not work if same webapp is deployed twice, so the
MethodHandle changes have been reverted and the test has been disabled
until a resolution to https://bugs.openjdk.java.net/browse/JDK-8244090.

- The DistributionTest has been changed to properly reproduce the issue.
There must be a class in the signature of the websocket endpoint which
is available to both webapps and the client must send and receive an echo
message to ensure that the onOpen MethodHandle is actually invoked.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Merge remote-tracking branch 'origin/jetty-10.0.x' into jetty-10.0.x-LinkageErrorInvestigation

lachlan-roberts committed:
fix WS DistributionTest incorrect test expectation

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts commented:
@sbordet this is ready for review. The WebSocket DistributionTest has been disabled on JDK 14 and now it is passing the build.
